### ISSUE TITLE
neovim stuck while pasting large text

### ISSUE THREAD

**Issue by charlielin** [neovim/neovim#19783]
### Neovim version (nvim -v)

v0.7.2

### Vim (not Nvim) behaves the same?

vim version: 8.2
yes, vim --clean
yes, vim -u DEFAULTS
**no, vim -u NONE**

### Operating system/version

Apple M1 Pro
macOS 12.5 Monterey(12.4 got this issue too)

### Terminal name/version

iTerm2 Build 3.4.16 & Terminal 2.12.7(445)

### $TERM environment variable

xterm-256color

### Installation

homebrew

### How to reproduce the issue

nvim --clean
:edit foo
press i into insert mode
<Command-v> paste a text about 100+ lines

### Expected behavior

paste 100+ lines correctly.

### Actual behavior

paste stuck maybe after 15 lines, but it's still ok to type alpha and number
while press \<esc>, shows `^[`
while press \<Backspace>, shows `^?`
while press \<C-v>, shows `^[99;5u`

**Comment by zeertzjq** [comment]
Does this happen for any 100+ lines text, or only some specific text?

**Comment by charlielin** [comment]
> 

Any 100+ lines

**Comment by beauwilliams** [comment]
I can reproduce it with the below linked text with no rc file (v0.7) 

https://pastebin.com/raw/7EX6ws3S

Steps to reproduce.

1. Copy/Paste the code in the above link

2. Open an existing file or new buffer

3. Enter insert mode (this is where the bug occurs)

4. Paste the code from the link from system clipboard (Cmd+V or Ctrl+V)

5. You will now be stuck in insert mode and there appears to be no way to exit.

**Comment by Kiran-Raj-Dev** [comment]
Hmm, looks like this issue is only reproducible on some terminals. I am unable to reproduce it in kitty.

**Comment by beauwilliams** [comment]
> Hmm, looks like this issue is only reproducible on some terminals. I am unable to reproduce it in kitty.

Good point. I've just tested on native osx terminal and the issue was not present.

However it is happening in iTerm2, my usual emulator. So for myself it appears to be an iTerm issue

**Comment by Kiran-Raj-Dev** [comment]
Not for me. I am on an M1 mac

**Comment by hazeled** [comment]
> Hmm, looks like this issue is only reproducible on some terminals. I am unable to reproduce it in kitty.

Yeah, I'm unable to reproduce on Ubuntu using alacritty

**Comment by PowerUser64** [comment]
I cannot reproduce this on neovim version `NVIM v0.8.0-dev-878-g02e9b5a8ee` (latest commit) or 0.7.0 with Alacritty on Manjaro. (Followed @beauwilliams' steps.)

**Comment by zeertzjq** [comment]
So it seems so far this has only been reproduced on macOS.

**Comment by PowerUser64** [comment]
It would be interesting to see if the problem persists when pasting from the `"+` register (new buffer -> `"+p`).

**Comment by Kiran-Raj-Dev** [comment]
Pasting from the unamed plus register is faster if anything :)

**Comment by clason** [comment]
I cannot reproduce this on macOS (with iTerm2).

**Comment by justinmk** [comment]
Pasting in insert-mode has really bad performance because Vim insists on treating it like user input, subject to all the auto-indent options, etc.

We should instead treat it like we treat normal-mode paste: turn off all that crap. Then the bottleneck will be your terminal (on macOS, Terminal.app is still horribly slow when it feeds text via bracketed-paste, whereas kitty and others will be near-instant).

**Comment by charlielin** [comment]
> If it also happens on Vim (which has a completely different bracketed-paste impl), it might be a shell configuration issue.
> 
> @charlielin @beauwilliams please try in a shell (without tmux/screen) without your config / environment:
> 
> ```
> env -i bash --norc --noprofile
> ```

@justinmk 
A new test in a shell without my config / enviroment:
#### vim version: 8.2
no, vim --clean
no, vim -u DEFAULTS
no, vim -u NONE
#### nvim v0.7.2
**yes, /opt/homebrew/bin/nvim --clean**

**Comment by justinmk** [comment]
Likely vim disabled its bracketed-paste handling (confirm with `:set termcap`) if `TERM=dumb`. Nvim always handles bracketed paste.

@charlielin are you able to reproduce the issue in Terminal.app or other terminals besides iTerm? 

@beauwilliams can you try `env -i bash --norc --noprofile` ?

**Comment by charlielin** [comment]
> @charlielin are you able to reproduce the issue in Terminal.app or other terminals besides iTerm?

yes, the behavior  in Terminal.app is as same as in iTerm.app

**Comment by justinmk** [comment]
please try the following:

```lua
phases = {}
vim.paste = (function(overridden)
 return function(lines, phase)
   table.insert(phases, phase)
   overridden(lines, phase)
 end
end)(vim.paste)
```

1. `:source` the above lua code.
2. check which phases were seen via `:lua =phases`

**Comment by justinmk** [comment]
> Enter insert mode (this is where the bug occurs)

This _only_ happens in insert-mode? Not in normal-mode?

**Comment by zeertzjq** [comment]
When stuck in bracketed paste mode it may be impossible to type a command.

**Comment by justinmk** [comment]
yeah let me update the code

**Comment by charlielin** [comment]
> 



> > Enter insert mode (this is where the bug occurs)
> 
> This _only_ happens in insert-mode? Not in normal-mode?

Yes, only happens in insert-mode with cmd-v to paste text. It's ok to paste text in normal-mode with `p`.

**Comment by zeertzjq** [comment]
Does it happen in Normal mode with Cmd-v?

**Comment by charlielin** [comment]
> please try the following:
> 
> ```lua
> vim.paste = (function(overridden)
>  return function(lines, phase)
>    local f = io.open('foo.txt','a')
>    f:write(tostring(phase))
>    io.close(f)
>    overridden(lines, phase)
>  end
> end)(vim.paste)
> ```
> 
> 1. `:write foo.lua`
> 2. `:source`
> 3. paste some text that reproduces this issue
> 4. what are the contents of `foo.txt` ?

@justinmk 
foo.txt:
11222222222222222222222222222222222222222222222222222222222222222222222222222222222222

**Comment by charlielin** [comment]
> Does it happen in Normal mode with Cmd-v?

OH YES, It also happens in Normal mode with Cmd-v

**Comment by zeertzjq** [comment]
"1" appearing twice without "3" appearing is also very strange. Maybe there is a data race?

**Comment by ad-chaos** [comment]
It just prints 13 for me. I am on a similar machine as them (M1 air, macos 12.5, terminal.app) and same nvim version of nvim (0.7.2)

**Comment by justinmk** [comment]
@ad-chaos are you seeing this bug? It should not happen if you are seeing "3"...

Can everyone try the latest development version? See the [Releases](https://github.com/neovim/neovim/releases/tag/nightly) page for pre-built downloads.

**Comment by ad-chaos** [comment]
I am not seeing seeing this bug, I tried it in kitty and terminal.app with the version they were running at and on head.

**Comment by charlielin** [comment]
> > foo.txt:
> > 11222222222222222222222222222222222222222222222222222222222222222222222222222222222222
> 
> Weird. Is the OS/terminal sending very tiny chunks? Or is Nvim hitting an infinite loop--does it keep appending "2" forever until you kill the `nvim` process? Does `vim` behave the same way (check its CPU usage in something like `top`)?

I think the out of `foo.txt` maybe is associated with the `esc` I press after stuck. Here is my test:
#### round 1
1. into insert-mode
2. paste 100+ lines text
3. nvim stuck and do not touch keyboard
4. close the iTerm.app
5. foo.txt: `1`

#### round 2
1. into insert-mode
2. paste 100+ lines text
3. nvim stuck and press `esc` once
4. close the iTerm.app
5. foo.txt: `122`

#### round 3
1. into insert-mode
2. paste 100+ lines text
3. nvim stuck and do press `esc` twice
4. close the iTerm.app
5. foo.txt: `12222`

#### round 4
1. into insert-mode
2. paste 100+ lines text
3. nvim stuck and do press `esc` thrice
4. close the iTerm.app
5. foo.txt: `1222222`

#### round 5
1. into insert-mode
2. paste 100+ lines text
3. nvim stuck and do not touch keyboard and wait for abount 60 seconds.
4. close the iTerm.app
5. foo.txt: `1`

**Comment by charlielin** [comment]
> @ad-chaos are you seeing this bug? It should not happen if you are seeing "3"...
> 
> Can everyone try the latest development version? See the [Releases](https://github.com/neovim/neovim/releases/tag/nightly) page for pre-built downloads.

I tried this pre-built release version. It's behavior is as same as `0.7.2`. Shows `1` or `1222222...`

**Comment by justinmk** [comment]
> #### round 1
> 1. into insert-mode
> 2. paste 100+ lines text
> 3. nvim stuck and do not touch keyboard
> 4. close the iTerm.app
> 5. foo.txt: `1`

Ok. Given that Vim behaves the same when its bracketed-paste support is enabled, this suggests a terminal or OS bug.

Only thing we can do is look for CTRL-C as a way to recover: https://github.com/neovim/neovim/issues/10865

**Comment by beauwilliams** [comment]
Sorry for delayed reply. Yes for me its only happening in insert mode. I've started to notice it more recently with a number of text snippets. I will try out the suggestion above now see if that makes a difference and let know if so

**Comment by beauwilliams** [comment]
This also happens to me in vim. Can safely say for my case its not specific to neovim

**Comment by charlielin** [comment]
Hello everyone, after I upgrade  operation system to version 12.5.1, the issue do not occur anymore.

**Comment by dundargoc** [comment]
I propose we close this then.

**Comment by jericopulvera** [comment]
Still happening to me. I'm using MacOS Ventura 13.1 and nvim 0.8.0 installed from bob version manager for neovim.

It does not happen on

```sh
env -i bash --norc --noprofile
```

I'm currently updating to MacOS 13.2.

<img width="747" alt="image" src="https://user-images.githubusercontent.com/23246308/217131870-0a4b2ff9-8e7b-410a-9c1c-6676e3c9a3f4.png">

**Comment by jericopulvera** [comment]
same issue on MacOS Ventura 13.2

**Comment by nevertoday** [comment]
my newest mac mini M2 have the problem too.

**Comment by justinmk** [comment]
> It does not happen on
> 
> ```shell
> env -i bash --norc --noprofile
> ```

hmm...

**Comment by griffinbaker12** [comment]
Can confirm that this is happening to me

**Comment by griffinbaker12** [comment]
Can confirm that this happens to me using zsh with iTerm on M1 mac, but does not happen when using a blank bash profile.

**Comment by kirilgorbachov** [comment]
+1 M2, iterm2, zsh

**Comment by YuraAka** [comment]
the problem is with environment variable TERM. If it's set to xterm-* (or maybe some other values), vim is constantly freezing during insert, but if you change TERM to 'linux' (for example), everything works fine, so I've come up with workaround: in .bashrc add `alias vim='TERM=linux vim'`.

**Comment by justinmk** [comment]
> I've come up with workaround: in .bashrc add `alias vim='TERM=linux vim'`.

That is a (very clunky and counterproductive) way to disable bracketed-paste.

**Comment by yeshwanth95** [comment]
This bug still occurs on WSL + neovim v0.10.1 on the Windows Terminal. Does anyone know a way to exit neovim when this happens?

**Comment by justinmk** [comment]
@yeshwanth95 there were major improvements to paste handling in Nvim HEAD, please try it: the [Releases](https://github.com/neovim/neovim/releases/tag/nightly) page has pre-built archives for Linux/Windows/macOS. Check `:version` to confirm the version.

**Comment by Kazz47** [comment]
Same, this issue still appears when pasting large files in insert mode on M2 Mac OSX, 15.2

NVIM v0.10.3

I wasn't able to reproduce it with the input @beauwilliams provided, but is reproducible with this: https://pastes.dev/tUeSnkoIR0

**Comment by axsaucedo** [comment]
I started seeing this issue only recently, one quick-fix I found is to update iterm2 settings.

Namely iterm2 > settings > advanced > scroll into the "Pasteboard section":

Then you will see an option: `"Delay in seconds between chunks when pasting normally"`

Update to the same value as the slow paste "0.125" and this solved the issue for me.

Not great but at least has solved it for mid-sized pastes which stopped working.

On my side it was working with `--clean` when I was on `v0.10.x` but when I upgraded to `v0.11.4` it doesn't work with `--clean` anymore.

If anyone finds a proper fix let me know.

**Comment by axsaucedo** [comment]
> I started seeing this issue only recently, one quick-fix I found is to update iterm2 settings.
> 
> Namely iterm2 > settings > advanced > scroll into the "Pasteboard section":
> 
> Then you will see an option: `"Delay in seconds between chunks when pasting normally"`
> 
> Update to the same value as the slow paste "0.125" and this solved the issue for me.
> 
> Not great but at least has solved it for mid-sized pastes which stopped working.
> 
> On my side it was working with `--clean` when I was on `v0.10.x` but when I upgraded to `v0.11.4` it doesn't work with `--clean` anymore.
> 
> If anyone finds a proper fix let me know.

Following up on this, updating to latest version of iterm seems has now fixed the issue, no need for workaround (`v3.6.1`). All good from my side! 

I'd suggest OP to re-test to see if has now been fixed as well.

**Comment by justinmk** [comment]
Thanks. General solution is still tracked in:

- https://github.com/neovim/neovim/issues/10865
